File: script/events.dart
class EventTarget {
    Map _listeners;
    Map _oneListeners;

    EventTarget() {
        _listeners = new Map();
        _oneListeners = new Map();
    }

    void fire(name, [data]) {
        if (_listeners.containsKey(name)) {
            for (Function listener in _listeners[name]) {
                Function.apply(listener, data);
            }
        }
        if (_oneListeners.containsKey(name)) {
            for (Function listener in _oneListeners[name]) {
                Function.apply(listener, data);
            }
            _oneListeners.remove(name);
        }
    }

    void on(name, Function listener) {
        _listeners.putIfAbsent(name, () => []);
        _listeners[name].add(listener);
    }

    void one(name, Function listener) {
        _oneListeners.putIfAbsent(name, () => []);
        _oneListeners[name].add(listener);
    }

    void off(name, Function listener) {
        if (_listeners.containsKey(name)) {
            _listeners[name].remove(listener);
            if (_listeners[name].isEmpty) {
                _listeners.remove(name);
            }
        }
        if (_oneListeners.containsKey(name)) {
            _oneListeners[name].remove(listener);
            if (_oneListeners[name].isEmpty) {
                _oneListeners.remove(name);
            }
        }
    }

}

File: src/events.js
export class EventTarget {
    constructor() {
        this.listeners = new Map();
        this.oneListeners = new Map();
    }

    fire(name, ...data) {
        if (this.listeners.has(name)) {
            this.listeners.get(name).forEach(listener => {
                listener(...data);
            });
        }
        if (this.oneListeners.has(name)) {
            this.oneListeners.get(name).forEach(listener => {
                listener(...data);
            });
            this.oneListeners.delete(name);
        }
    }

    on(name, listener) {
        if (!this.listeners.has(name)) {
            this.listeners.set(name, new Set());
        }
        this.listeners.get(name).add(listener);
    }

    one(name, listener) {
        if (!this.oneListeners.has(name)) {
            this.oneListeners.set(name, new Set());
        }
        this.oneListeners.get(name).add(listener);
    }

    off(name, listener) {
        if (this.listeners.has(name)) {
            this.listeners.get(name).delete(listener);
            if (this.listeners.get(name).size() === 0) {
                this.listeners.delete(name);
            }
        }
        if (this.oneListeners.has(name)) {
            this.oneListeners.get(name).delete(listener);
            if (this.oneListeners.get(name).size() === 0) {
                this.oneListeners.delete(name);
            }
        }
    }
};
